### PR TITLE
enhancement: different library used

### LINKED ISSUE #2: Programmatic CSR Generation
Implement programmatic CSR export using node-webcrypto-p11 and @peculiar/x509 libraries.

See issue comment for reference: https://github.com/PeculiarVentures/pkcs11js/issues/80#issuecomment-845461501.

### LINKED ISSUE #3: Refactor generated key storage

### LINKED ISSUE #5: Document functions
Please take a look at the [JSDoc](https://jsdoc.app/) specification.

### PR DESCRIPTION
**Pull request by atasen19**

### CODE REVIEW

File: scripts/generate.js [hunk 1]
```diff
@@ -1,11 +1,20 @@
...
+  publicExponent: new Uint8Array([1, 0, 1]),
+  modulusLength: 2048,
+  token: true,
+  sensitive: true
+}
 
-(async function () {
-  await generate();
+const { keys } = await generateKeyPair(alg)
+console.warn('privKey', keys.privateKey)
```
Comment (mcaneris) [inline scripts/generate.js]: we shouldn't print the key here.

File: scripts/generate.js [hunk 2]
```diff
@@ -1,11 +1,20 @@
...
 
-  execSync(
-    "sudo openssl req -nodes -new -days 365 -sha256 -config engine.conf -engine pkcs11 -keyform engine -key slot_81 -out pandora.csr"
-  );
-})();
+await generateCSR(alg, keys)
+
+await sign(alg, keys, 'data_to_be_signed')
```
Comment (mcaneris) [inline scripts/generate.js]: no signing here. this is a generation script.

File: scripts/generate.js [hunk 3]
```diff
@@ -1,11 +1,20 @@
...
-(async function () {
-  await generate();
+const { keys } = await generateKeyPair(alg)
+console.warn('privKey', keys.privateKey)
 
-  execSync(
-    "sudo openssl req -nodes -new -days 365 -sha256 -config engine.conf -engine pkcs11 -keyform engine -key slot_81 -out pandora.csr"
-  );
-})();
+await generateCSR(alg, keys)
```
Comment (mcaneris) [inline scripts/generate.js]: output CSR to local file.

File: src/csr.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    const csr = await x509.Pkcs10CertificateRequestGenerator.create({
+      name: 'CN=xyzteknoloji.com, C=TR, ST=Istanbul, L=Kozyatagi',
+      keys,
+      signingAlgorithm: alg,
+      attributes: [
+        new x509.ChallengePasswordAttribute('123456')
+      ]
+    }, crypto)
+
+    console.log(csr.toString('pem'))
```
Comment (mcaneris) [inline src/csr.js]: do not log. return the CSR.

File: src/generator.js [hunk 5]
```diff
@@ -1,56 +1,32 @@
...
+    // They can be cleaned in this way.
+    await crypto.keyStorage.clear()
+    await crypto.certStorage.clear()
+
+    const keys = await crypto.subtle.generateKey(alg, false, ['sign', 'verify'])
+
+    const allKeys = await crypto.keyStorage.keys()
+    for (const index of allKeys) {
+      console.log('all keys', index)
+    }
```
Comment (mcaneris) [inline src/generator.js]: do not log

File: src/generator.js [hunk 6]
```diff
@@ -1,56 +1,32 @@
...
+      name: 'TÜBİTAK NHSM',
+      slot: 0,
+      pin: '123456',
+      readWrite: true
+    })
+
+    // Once keys created they stored unless any bad action.
+    // They can be cleaned in this way.
+    await crypto.keyStorage.clear()
+    await crypto.certStorage.clear()
```
Comment (mcaneris) [inline src/generator.js]: do not clear

File: src/generator.js [hunk 7]
```diff
@@ -1,56 +1,32 @@
...
-  { type: pkcs11js.CKA_TOKEN, value: false },
-  { type: pkcs11js.CKA_LABEL, value: "rsa_private_key" },
-  { type: pkcs11js.CKA_SIGN, value: true },
-];
-
-const pkcs11 = new pkcs11js.PKCS11();
-pkcs11.load("/usr/local/lib/libdirakp11-64.so");
-
-async function generate() {
+async function generateKeyPair (alg) {
```
Comment (mcaneris) [inline src/generator.js]: receive slot to create as argument

File: src/generator.js [hunk 8]
```diff
@@ -1,56 +1,32 @@
...
+    await crypto.certStorage.clear()
+
+    const keys = await crypto.subtle.generateKey(alg, false, ['sign', 'verify'])
+
+    const allKeys = await crypto.keyStorage.keys()
+    for (const index of allKeys) {
+      console.log('all keys', index)
+    }
+
+    return { alg, keys, crypto }
```
Comment (mcaneris) [inline src/generator.js]: why return anything except keys?

File: src/signer.js [hunk 9]
```diff
@@ -1,42 +1,18 @@
...
-
-    pkcs11.C_CloseSession(session);
-  } catch (e) {
-    console.error(e);
-  } finally {
-    pkcs11.C_Finalize();
-  }
+import { Crypto } from 'node-webcrypto-p11'
+
+async function sign (alg, keys, data) {
```
Comment (mcaneris) [inline src/signer.js]: receive slot to read the key from as argument.

File: src/signer.js [hunk 10]
```diff
@@ -1,42 +1,18 @@
...
+  const crypto = new Crypto({
+    library: '/usr/local/lib/libdirakp11-64.so',
+    name: 'TÜBİTAK NHSM',
+    slot: 0,
+    pin: '123456',
+    readWrite: true
+  })
+
+  const signature = await crypto.subtle.sign(alg, keys.privateKey, Buffer.from(data))
+  console.log(`Signature: ${signature}`)
```
Comment (mcaneris) [inline src/signer.js]: do not log.

File: src/signer.js [hunk 11]
```diff
@@ -1,42 +1,18 @@
...
+  const ok = await crypto.subtle.verify(alg, keys.publicKey, signature, Buffer.from(data))
+  console.log(`Verification: ${ok}`)
```
Comment (mcaneris) [inline src/signer.js]: do not log.

Comment (mcaneris) [reply src/signer.js]: return signature.

**Review by mcaneris — CHANGES_REQUESTED** [review]

File: src/signer.js (hunk 9 above)
Comment (atasen19) [reply src/signer.js]: I couldn't figure it out how to receive the key from with requested type / or there is no way to do it according to the library's documentation, which means we have to take it as an input.

File: src/signer.js [hunk 12]
```diff
@@ -1,42 +1,18 @@
...
-    pkcs11.C_Finalize();
-  }
+import { Crypto } from 'node-webcrypto-p11'
+
+async function sign (alg, keys, data) {
+  const crypto = new Crypto({
+    library: '/usr/local/lib/libdirakp11-64.so',
+    name: 'TÜBİTAK NHSM',
+    slot: 0,
+    pin: '123456',
```
Comment (mcaneris) [inline src/signer.js]: read these values from env

**Review by mcaneris — APPROVED** [review]

**Pull request merged**